README
NAME
    Perl::Critic::Policy::logicLAB::RequireSheBang - simple policy for
    keeping your shebang line uniform

AFFILIATION
    This policy is a policy in the Perl::Critic::logicLAB distribution. The
    policy is themed: logiclab.

VERSION
    This documentation describes version 0.03.

DESCRIPTION
    This policy is intended in guarding your use of the shebang line. It
    assists in making sure that your shebang line adheres to certain
    formats.

    The default format is

        #!/usr/local/bin/perl

    You can however specify another or define your own in the configuration
    of the policy.

    NB this policy does currently not warn about missing shebang lines, it
    only checks shebang lines encountered.

CONFIGURATION AND ENVIRONMENT
    This policy allow you to configure the contents of the shebang lines you
    want to allow using "formats".

  formats
        [logicLAB::RequireSheBang]
        formats = #!/usr/local/bin/perl || #!/usr/bin/perl || #!perl || #!env perl

    Since the default shebang line enforced by the policy is:

        #!/usr/local/bin/perl

    Please note that if you however what to extend the pattern, you also
    have to specify was is normally the default pattern since configuration
    overwrites the default even for extensions.

    This mean that if you want to allow:

        #!/usr/local/bin/perl

        #!/usr/local/bin/perl -w
    
        #!/usr/local/bin/perl -wT

    Your format should look like the following:

        [logicLAB::RequireSheBang]
        formats = #!/usr/local/bin/perl || #!/usr/local/bin/perl -w || #!/usr/local/bin/perl -wT

  debug
    Optionally and for development purposes I have added a debug flag. This
    can be set in your Perl::Critic configuration file as follows:

        [logicLAB::RequireSheBang]
        debug = 1

    This enables more explicit output on what is going on during the actual
    processing of the policy.

DEPENDENCIES AND REQUIREMENTS
    *   Perl::Critic

    *   Perl::Critic::Utils

    *   Readonly

    *   Test::More

    *   Test::Perl::Critic

    *   List::MoreUtils

INCOMPATIBILITIES
    This distribution has no known incompatibilities.

BUGS AND LIMITATIONS
    The distribution has now known bugs or limitations. It locates shebang
    lines through out the source code, not limiting itself to the first
    line. This might however change in the future, but will propably be made
    configurable is possible.

BUG REPORTING
    Please use Requets Tracker for bug reporting:

        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-RequireSheBang

TEST AND QUALITY
    The following policies have been disabled for this distribution

    *   Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma

    *   Perl::Critic::Policy::NamingConventions::Capitalization

    *   Data::Dumper

    See also t/perlcriticrc

  TEST COVERAGE
    Coverage test executed the following way, the coverage report is based
    on the version described in this documentation (see "VERSION").

        ./Build testcover

            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            File                           stmt   bran   cond    sub    pod   time  total
            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            ...ogicLAB/RequireSheBang.pm   70.4   64.3   44.4  100.0  100.0  100.0   72.1
            Total                          70.4   64.3   44.4  100.0  100.0  100.0   72.1
            ---------------------------- ------ ------ ------ ------ ------ ------ ------

SEE ALSO
    *   Perl::Critic

    *   <http://perldoc.perl.org/perlrun.html>

    *   <http://logiclab.jira.com/wiki/display/OPEN/Development#Development-
        MakeyourComponentsEnvironmentAgnostic>

    *   <http://logiclab.jira.com/wiki/display/PCPLRSB/Home>

    *   <http://logiclab.jira.com/wiki/display/PCLL/Home>

AUTHOR
    *   Jonas B. Nielsen, jonasbn "<jonasbn@cpan.org>"

ACKNOWLEDGEMENT
    *   Erik Johansen (uniejo), feedback to version 0.01

LICENSE AND COPYRIGHT
    Copyright (c) 2011 Jonas B. Nielsen, jonasbn. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Changes file for Perl::Critic::Policy::logicLAB::RequireSheBang

$Id: Changes 8192 2013-08-02 08:40:15Z jonasbn $

Project change log:

http://logiclab.jira.com/browse/PCPLRSB#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.04 2013-08-02

- Maintenance release, update not necessary

- Adding a Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [PCPLRSB-10]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec
  
  https://logiclab.jira.com/browse/PCPLRSB-10
  
- Minor improvement/correction to the central regular expression [PCPLRSB-9
  
  https://logiclab.jira.com/browse/PCPLRSB-9]

- Added a debugging flag [PCPLRSB-8]
  
  https://logiclab.jira.com/browse/PCPLRSB-8


0.03 2011-06-06

- Bug fix release, update recommended

- Serious bug fix to parsing of shebang lines [PCPLRSB-6]
  http://logiclab.jira.com/browse/PCPLRSB-6


0.02 2011-05-16

- Bug fix release, Update recommended

- Minor corrections to documentation [PCPLRSB-3]
  http://logiclab.jira.com/browse/PCPLRSB-3
  
  Theme changed to logiclab from maintenance according to the guidelines, I
  had previously overlooked this:
  http://search.cpan.org/dist/Perl-Critic/lib/Perl/Critic/DEVELOPER.pod#Themes

- Added Changes (this file)


0.01 2011.05.03

- Initial release


Build.PL
# Build.PL for Perl::Critic::Policy::logicLAB::RequireSheBang

# $Id: Build.PL 7636 2011-05-01 15:24:04Z jonasbn $

use strict;
use Module::Build;

my $build = Module::Build->new(
    dist_author          => 'Jonas B. Nielsen, <jonasbn@cpan.org>',
    module_name          => 'Perl::Critic::Policy::logicLAB::RequireSheBang',
    license              => 'artistic',
    recursive_test_files => 0,
    create_makefile_pl   => 'traditional',
    create_readme        => 1,
    dist_abstract        => q[simple policy asserting she bang line],
    requires => {
        'Perl::Critic'        => '0',
        'Perl::Critic::Utils' => '0',
        'Readonly'            => '0',
        'Test::More'          => '0',
        'Test::Perl::Critic'  => '0',
        'List::MoreUtils'     => '0',
    },
    meta_merge => {
        resources => {
            homepage => 'http://logiclab.jira.com/wiki/display/PCPLRSB/Home',
            bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-Require-SheBang',
            repository => 'http://logicLAB.jira.com/svn/PCPLRSB',
            ProjectChangelog =>
                'http://logiclab.jira.com/browse/PCPLRSB#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel',
            ProjectSite => 'http://logicLAB.jira.com/browse/PCPLRSB',
            ProjectWiki => 'http://logiclab.jira.com/wiki/display/PCPLRSB/Home',
        }
    },
);

$build->create_build_script();

1;


lib/Perl/Critic/Policy/logicLAB/RequireSheBang.pm
package Perl::Critic::Policy::logicLAB::RequireSheBang;

# $Id: RequireSheBang.pm 8192 2013-08-02 08:40:15Z jonasbn $

use strict;
use warnings;
use base 'Perl::Critic::Policy';
use Perl::Critic::Utils qw{ $SEVERITY_MEDIUM :booleans };
use List::MoreUtils qw(none);
use Data::Dumper;

$Data::Dumper::Useqq = 1;

our $VERSION = '0.04';

Readonly::Scalar my $EXPL  => q{she-bang line should adhere to requirement};
Readonly::Scalar my $DEBUG => q{DEBUG logicLAB::RequireSheBang};

use constant default_severity     => $SEVERITY_MEDIUM;
use constant default_themes       => qw(logiclab);
use constant supported_parameters => qw(formats debug);

sub applies_to {
    return (
        qw(
            PPI::Token::Comment
            )
    );
}

sub violates {
    my ( $self, $elem, $doc ) = @_;
    
    if ( $self->{debug} ) {
        print {*STDERR} "$DEBUG: we got element:\n";
        print {*STDERR} Dumper $elem;
    }

    my ( $shebang, $cli ) = $elem =~ m{
            \A  #beginning of string
            (\#!) #actual she-bang
            #([^\r\n]*?) #the path and possible flags, note the space character
            ([/\-\w ]+?) #the path and possible flags, note the space character
            \s* #possible left over whitespace (PPI?)
            \Z #indication of end of string to assist above capture
    }xsm;

    if ($cli) {
        $cli =~ s{
            \s+ #one or more whitespace character, PCPLRSB-9 / http://logiclab.jira.com/browse/PCPLRSB-9
            $ #end of string
        }{}xsm;
    }

    if ( $self->{debug} && $shebang && $cli ) {
        print {*STDERR} "$DEBUG: we got a shebang line:\n";
        print {*STDERR} '>' . $shebang . $cli . "<\n";

        print {*STDERR} "$DEBUG: comparing against formats:\n";
        print {*STDERR} @{ $self->{_formats} };
        print {*STDERR} "\n";

    } elsif ( $self->{debug} ) {
        print {*STDERR} "$DEBUG: not a shebang, ignoring...\n";
    }

    if ( $shebang && none { ( $shebang . $cli ) eq $_ }
        @{ $self->{_formats} } )
    {

        if ( $self->{debug} ) {
            print {*STDERR} "$DEBUG: we got a violation:\n";
            print {*STDERR} '>' . $shebang . $cli . "<\n";
        }

        return $self->violation(
            q{she-bang line not conforming with requirement},
            $EXPL, $elem );
    }

    return;
}

sub initialize_if_enabled {
    my ( $self, $config ) = @_;

    #Formats:
    #Setting the default
    $self->{_formats} = [ ('#!/usr/local/bin/perl') ];

    #fetching configured formats
    my $formats = $config->get('formats');

    #parsing configured formats, see also _parse_formats
    if ($formats) {
        $self->{_formats} = $self->_parse_formats($formats);
    }

    #Debug
    #Setting the default
    $self->{debug} = $config->get('debug');

    return $TRUE;
}

sub _parse_formats {
    my ( $self, $config_string ) = @_;

    my @formats = split m{ \s* [||]+ \s* }xsm, $config_string;

    return \@formats;
}

1;

__END__

=pod

=head1 NAME

Perl::Critic::Policy::logicLAB::RequireSheBang - simple policy for keeping your shebang line uniform

=head1 AFFILIATION 

This policy is a policy in the Perl::Critic::logicLAB distribution. The policy
is themed: logiclab.

=head1 VERSION

This documentation describes version 0.03.

=head1 DESCRIPTION

This policy is intended in guarding your use of the shebang line. It assists
in making sure that your shebang line adheres to certain formats.

The default format is

    #!/usr/local/bin/perl
    
You can however specify another or define your own in the configuration of the
policy.

B<NB> this policy does currently not warn about missing shebang lines, it only
checks shebang lines encountered.

=head1 CONFIGURATION AND ENVIRONMENT

This policy allow you to configure the contents of the shebang lines you 
want to allow using L</formats>.

=head2 formats

    [logicLAB::RequireSheBang]
    formats = #!/usr/local/bin/perl || #!/usr/bin/perl || #!perl || #!env perl

Since the default shebang line enforced by the policy is:

    #!/usr/local/bin/perl
    
Please note that if you however what to extend the pattern, you also have 
to specify was is normally the default pattern since configuration 
overwrites the default even for extensions.

This mean that if you want to allow:

    #!/usr/local/bin/perl

    #!/usr/local/bin/perl -w
    
    #!/usr/local/bin/perl -wT
        
Your format should look like the following:

    [logicLAB::RequireSheBang]
    formats = #!/usr/local/bin/perl || #!/usr/local/bin/perl -w || #!/usr/local/bin/perl -wT

=head2 debug

Optionally and for development purposes I have added a debug flag. This can be set in 
your L<Perl::Critic> configuration file as follows:

    [logicLAB::RequireSheBang]
    debug = 1

This enables more explicit output on what is going on during the actual processing of 
the policy.

=head1 DEPENDENCIES AND REQUIREMENTS

=over

=item * L<Perl::Critic>

=item * L<Perl::Critic::Utils>

=item * L<Readonly>

=item * L<Test::More>

=item * L<Test::Perl::Critic>

=item * L<List::MoreUtils>

=back

=head1 INCOMPATIBILITIES

This distribution has no known incompatibilities.

=head1 BUGS AND LIMITATIONS

The distribution has now known bugs or limitations. It locates shebang lines 
through out the source code, not limiting itself to the first line. This might 
however change in the future, but will propably be made configurable is possible.

=head1 BUG REPORTING

Please use Requets Tracker for bug reporting:

    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-RequireSheBang

=head1 TEST AND QUALITY

The following policies have been disabled for this distribution

=over

=item * L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma>

=item * L<Perl::Critic::Policy::NamingConventions::Capitalization>

=item * L<Data::Dumper>

=back

See also F<t/perlcriticrc>

=head2 TEST COVERAGE

Coverage test executed the following way, the coverage report is based on the 
version described in this documentation (see L</VERSION>).

    ./Build testcover

	---------------------------- ------ ------ ------ ------ ------ ------ ------
	File                           stmt   bran   cond    sub    pod   time  total
	---------------------------- ------ ------ ------ ------ ------ ------ ------
	...ogicLAB/RequireSheBang.pm   70.4   64.3   44.4  100.0  100.0  100.0   72.1
	Total                          70.4   64.3   44.4  100.0  100.0  100.0   72.1
	---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 SEE ALSO

=over

=item * L<Perl::Critic>

=item * L<http://perldoc.perl.org/perlrun.html>

=item * L<http://logiclab.jira.com/wiki/display/OPEN/Development#Development-MakeyourComponentsEnvironmentAgnostic>

=item * L<http://logiclab.jira.com/wiki/display/PCPLRSB/Home>

=item * L<http://logiclab.jira.com/wiki/display/PCLL/Home>

=back

=head1 AUTHOR

=over

=item * Jonas B. Nielsen, jonasbn C<< <jonasbn@cpan.org> >>

=back

=head1 ACKNOWLEDGEMENT

=over

=item * Erik Johansen (uniejo), feedback to version 0.01

=back 

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011 Jonas B. Nielsen, jonasbn. All rights reserved.

This program is free software; you can redistribute it and/or modify it under the 
same terms as Perl itself.

=cut
